### PR TITLE
Split TBAA for array buffers

### PR DESCRIPTION
**Pull request by Keno**
Array buffers with boxed and unboxed data do not alias. Expose this info to LLVM.

### CODE REVIEW

**Review by yuyichao — APPROVED** [review]

File: src/codegen.cpp [hunk 1]
```diff
@@ -2827,7 +2828,8 @@ static bool emit_builtin_call(jl_cgval_t *ret, jl_value_t *f, jl_value_t **args,
                         *ret = ghostValue(ety);
                     }
                     else {
-                        *ret = typed_load(emit_arrayptr(ary, args[1], ctx), idx, ety, ctx, tbaa_arraybuf);
+                        *ret = typed_load(emit_arrayptr(ary, args[1], ctx), idx, ety, ctx,
```
Comment (yuyichao) [inline src/codegen.cpp]: seems that you can have a `bool isboxed` a few lines above for this one too.

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21262#issuecomment-291245974) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/e5438b3fbbc3a1b1d8ca8be03276b499f6676f06/9d9525f_vs_daefda4/report.md). cc @jrevels

**Comment by Keno** [comment]
Well, those performance results are moderately disappointing. I'll take a look.

**Comment by Keno** [comment]
I'm not able to reproduce these performance regressions:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21262#issuecomment-292063453) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/aa24849eadfe24282b704164990e55d1479802dd/853bb75_vs_9ed7d82/report.md). cc @jrevels

**Comment by Keno** [comment]
Nanosoldier results look like noise. Few of them are the same between the two runs, and one that is got faster in the second run. Merging.

File: src/codegen.cpp (hunk 1 above)
Comment (tkelman) [inline src/codegen.cpp]: did not get addressed?

Comment (Keno) [inline src/codegen.cpp]: Just a small cleanup, will get it in separately.

**Pull request merged**